src/models/gamma_poisson_model.py

```diff
@@ -247,116 +249,319 @@ def _impression_count(self, h):
+        return h
 
-    def _fit_histogram_chi2_distance(self, h):
+    def _histogram_with_kplus_bucket(self, point):
+        """Computes a histogram whose last bucket is k+ reach.
+
+        Args:
+          point:  A ReachPoint for which the histogram is to be
+            computed.
+        Returns:
+          A list h[] where h[i] is the number of people reached
+          exactly i+1 times, except for h[-1], which records the
+          number of people reached len(h) or more times.
+        """
+        # A partial histogram of people who were reached exactly k times.
+        h = self._histogram_from_reach_point(point)
+
+        # The number of people who were reach len(h) or more times.  These
+        # people are not counted in the histogram h.
+        kplus = len(h) + 1
+        kplus_reach = point.reach(kplus)
+
+        # Append kplus_reach to h.  So, the last value of h is treated
+        # specially.
+        h.append(kplus_reach)
+        h = np.array(h)
+        return h
+
+    def _chi2_distance_with_one_plus_reach(
+        self, h, I, N, alpha, beta, one_plus_reach_weight=1.0
+    ):
+        """Returns distance between actual and expected histograms.
+
+        Computes the metric
+
+            chi2(h, hbar) = \sum_i (h[i] - hbar[i])^2 / hbar[i] +
+               w * (\sum_i (h[i] - hbar[i]))^2
+
+        where
+            hbar is the histogram that would be expected given parameters
+               I, N, alpha, beta
+            w is the one_plus_reach_weight that is used to determine how
+               much to weight an exact match of 1+ reach.
+
+        The first part of this objective formula is the chi-squared statistic
+        used for determining whether two sets of categorical samples are drawn
+        from the same underlying distribution.  Therefore, if the value returned
+        by this function is relatively low (say < 20), then there is reason to
+        believe that the Gamma-Poisson distribution adequately fits the data.
+
+        Args:
+          h: Actual frequency histogram.  h[i] is the number of people
+             reached exactly i+1 times, except for h[-1] which is the
+             number of people reached len(h) or more times.
+          I: Total number of impressions purchased.
+          N: Total number of unique users.
+          alpha:  Parameter of the Gamma-Poisson distribution.
+          beta:  Second parameter of the Gamma-Poisson distribution.
+          one_plus_reach_weight:  Weight to be given to the squared
+            difference in 1+ reach values for the actual and estimated
+            histograms.
+        Returns:
+          chi^2 distance between the actual and expected histograms, plus
+          a term for weighting the difference in 1+ reach.
+        """
+        # Estimate total number of potential impressions
+        Imax = self._expected_impressions(N, alpha, beta)
+        if Imax <= I:
+            return np.sum(np.array(h) ** 2)
```

Comment:
Do you know in what scenarios we will have `Imax <= I`? 

In addition, I wonder the huge flat loss surface you mentioned in the slides is caused by these two lines which always outputs `np.sum(np.array(h) ** 2)`.

Reply:
I is the number of impressions that were actually served on behalf of the campaign.  Imax is the total number of impressions that could have been served if all impressions belonging to all users were purchased.  In principle, it should never be the case that Imax < I, because that would mean that more impressions were served than were available to  be served.  Since Imax = N * (1 + alpha * beta), it can be less than I if both alpha and beta are sufficiently small.  So, in the graphs of the loss surface, one sees a high wall along the borders of the surface, where alpha or beta or both are very small.  This high wall is the region where Imax < I.  The huge flat loss surface occurs in the interior, where I < Imax.

Reply:
Got it. Thank you so much for explaining this to me.